cpp/src/column/column_device_view.cu

```diff
@@ -45,51 +44,72 @@ column_device_view::column_device_view( column_view source, ptrdiff_t h_ptr, ptr
-    CUDF_FAIL("Columns with grand-children are not currently supported.");
-  }
   auto deleter = [](column_device_view* v) { v->destroy(); };
   std::unique_ptr<column_device_view, decltype(deleter)> p{
       new column_device_view(source), deleter};
 
   if( num_children > 0 )
   {
-    // create device memory for the children
-    RMM_TRY(RMM_ALLOC(&p->d_children, sizeof(column_device_view)*num_children, stream));
-    // build the children into CPU memory first
-    std::vector<uint8_t> buffer(sizeof(column_device_view)*num_children);
-    auto h_ptr = buffer.data();
+    // First calculate the size of memory needed to hold the
+    // child columns. This is done by calling extent()
+    // for each of the children.
+    size_type size_bytes = 0;
+    for( size_type idx=0; idx < num_children; ++idx )
+      size_bytes += extent(source.child(idx));
+    // A buffer of CPU memory is allocated to hold the column_device_view
+    // objects. Once filled, the CPU memory is then copied to device memory
+    // at the d_children member pointer.
+    std::vector<int8_t> h_buffer(size_bytes);
+    column_device_view* h_column = reinterpret_cast<column_device_view*>(h_buffer.data());
+    // Each column_device_view instance may have child objects that
+    // require setting some internal device pointers before being copied
+    // from CPU to device.
+    RMM_TRY(RMM_ALLOC(&p->d_children, size_bytes, stream));
+    column_device_view* d_column = p->d_children;
+    // The beginning of the memory must be the fixed-sized column_device_view
+    // struct objects in order for d_children to be used as an array. Therefore,
+    // any child data is assigned to the end of this array.
+    int8_t* h_end = (int8_t*)(h_column + num_children);
+    int8_t* d_end = (int8_t*)(d_column + num_children);
     for( size_type idx=0; idx < num_children; ++idx )
     {
       // create device-view from view
-      column_device_view child(source.child(idx));
+      auto child = source.child(idx);
       // copy child into buffer
-      memcpy(h_ptr, &child, sizeof(column_device_view));
+      new(h_column) column_device_view(child,(ptrdiff_t)h_end,(ptrdiff_t)d_end);
       // point to the next array slot
-      h_ptr += sizeof(column_device_view);
+      h_column++; // point to memory slot for the next child
+      // update the pointers for holding this child column's child data
+      auto col_child_data_size = extent(child) - sizeof(child);
+      h_end += col_child_data_size;
+      d_end += col_child_data_size;
     }
-    // copy the CPU memory with the children into device memory
-    CUDA_TRY(cudaMemcpyAsync(p->d_children, buffer.data(), num_children*sizeof(column_device_view),
+    // copy the CPU memory with all the children into device memory
+    CUDA_TRY(cudaMemcpyAsync(p->d_children, h_buffer.data(), size_bytes,
                               cudaMemcpyHostToDevice, stream));
     p->_num_children = num_children;
     cudaStreamSynchronize(stream);
```

Comment:
Is this sync necessary? If they've passed a non-default stream, the caller is responsible for syncing it before accessing the view using a different non-default stream, right?

Reply:
This is because the `h_buffer` is about go out of scope and be freed. Maybe this is overly cautious but if the copy is asynchronous the host pointer will be destroyed right after the sync.

Reply:
OK, makes sense.